metadata/src/main/java/com/redhat/lightblue/metadata/mongo/MetadataCache.java

```diff
@@ -0,0 +1,180 @@
+
+ This file is part of lightblue.
+
+ This program is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ This program is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.redhat.lightblue.metadata.mongo;
+
+import java.util.Map;
+import java.util.HashMap;
+
+import java.lang.ref.WeakReference;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.mongodb.DBCollection;
+import com.mongodb.BasicDBObject;
+import com.mongodb.DBObject;
+import com.mongodb.WriteResult;
+
+import com.redhat.lightblue.EntityVersion;
+
+import com.redhat.lightblue.metadata.EntityMetadata;
+
+public class MetadataCache {
+    
+    private static Logger LOGGER=LoggerFactory.getLogger(MetadataCache.class);
+
+    private static final String LITERAL_COLL_VER="collectionVersion";
+
+    /**
+     * This is the collection version number we expect to see in the
+     * database. If this doesn't match the value in db, someone
+     * updated metadata, we refresh
+     */
+    private long expectedCollectionVersion;
+
+    /**
+     * Last time we retrieved collection version
+     */
+    private volatile long lastVersionLookupTime=0l;
+
+    /**
+     * The collection version lookup period
+     */
+    private long versionLookupPeriodMsecs=10l*1000l;
+
+    /**
+     * Cache clear period
+     */
+    private long cacheTTLMsecs=10l*60l*1000l;
+
+    /**
+     * Last time we refreshed cache
+     */
+    private volatile long lastCacheRefreshTime=0l;
+    
+    private final Map<EntityVersion,WeakReference<EntityMetadata>> cache=new HashMap<>();
+
+    /**
+     * Sets cache parameters. If null is passed, that parameter is not changed.
+     */
+    public void setCacheParams(Long versionLookupPeriodMsecs,
+                               Long cacheTTLMsecs) {
+        if(versionLookupPeriodMsecs!=null)
+            this.versionLookupPeriodMsecs=versionLookupPeriodMsecs;
+        if(cacheTTLMsecs!=null)
+            this.cacheTTLMsecs=cacheTTLMsecs;
+    }
+    
+    public EntityMetadata lookup(DBCollection collection,String entityName,String version) {
+        EntityMetadata md;
+        EntityVersion v=new EntityVersion(entityName,version);
+        WeakReference<EntityMetadata> ref=cache.get(v);
+        if(ref!=null)
+            md=ref.get();
+        else
+            md=null;
+
+        long now=System.currentTimeMillis();
+        if(lastCacheRefreshTime+cacheTTLMsecs<now)
+            fullRefresh(collection,now);
```

Comment:
Why isn't full refresh being done before getting something from the cache?  At this point we'll have stale metadata, at least it's at the edge of the TTL for the cache.